Highlight relevant toolbar button when in a matching text node

chandlerprall opened the pull request:
### Summary

Opening as a draft to ensure CI is happy before reviews

* properly linked the toolbar buttons with their markdown AST nodes so the buttons will highlight as the user navigates the markdown
* added `aria-pressed` to the toolbar buttons
* created an initial markdown cypress spec, added test cases for the button<->node linkage

I wanted to take this further and make the buttons highlight for nested nodes (e.g. italic in bold), but the button actions don't function in that environment so I think it's better to avoid handling element nesting when determining the button state.

#### Before

![markdown](https://user-images.githubusercontent.com/313125/165146115-2f1f11f0-41fb-4a88-946c-5ac9e28eeb65.gif)

#### After

![markdown-after](https://user-images.githubusercontent.com/313125/165146155-fa97456b-f6ae-45d5-b81c-35dac88c7abc.gif)

### Checklist

~- [ ] Checked in both **light and dark** modes~
- [x] Checked in **mobile**
- [x] Checked in **Chrome**, **Safari**, **Edge**, and **Firefox**
~- [ ] Props have proper **autodocs** and **[playground toggles](https://github.com/elastic/eui/blob/main/wiki/documentation-guidelines.md#adding-playground-toggles)**~
~- [ ] Added **[documentation](https://github.com/elastic/eui/blob/main/wiki/documentation-guidelines.md)**~
~- [ ] Checked **[Code Sandbox](https://codesandbox.io/)** works for any docs examples~
- [x] Added or updated **[jest](https://github.com/elastic/eui/blob/main/wiki/testing.md) and [cypress](https://github.com/elastic/eui/blob/main/wiki/cypress-testing.md) tests**
- [x] Checked for **breaking changes** and labeled appropriately
- [x] Checked for **accessibility** including keyboard-only and screenreader modes
~- [ ] Updated the **[Figma](https://www.figma.com/community/file/964536385682658129)** library counterpart~
- [ ] A **[changelog](https://github.com/elastic/eui/blob/main/wiki/documentation-guidelines.md#changelog)** entry exists and is marked appropriately

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_5840/

chandlerprall commented:
@1Copenut I requested your review specifically for the added `aria-pressed` attribute, no hurry on taking a look

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_5840/

constancecchen commented on src/components/markdown_editor/markdown_editor_toolbar.tsx:
```diff
@@ -88,6 +89,42 @@ const quoteCodeLinkButtons = [
+  label: string;
+  icon: IconType;
+}
+export const formatSelectableButton = ({
```
Dumb question: If this takes an obj of props and returns JSX, why make this a vanilla JS function instead of an actual React component? e.g.,

```jsx
export const EuiToolbarButton = ({ selectedNode, handleMdButtonClick, ... }) => {
 // etc.
}
```

chandlerprall replied:
Not at all a dumb question, that was a great suggestion; converted the function to a component and pushed it up.

thompsongl commented:
jenkins test this

constancecchen commented:
Super brief QA testing notes:

1. ~it looks like stacking bold and italics doesn't highlight both - not sure if that matters?~ EDIT: just saw your PR description sorry, that totally makes sense re: behavior. I'm good with it

2. Quotes only appear to highlight when the first word of the quote, and not when clicking anywhere in the quote without selection
![screencap](https://user-images.githubusercontent.com/549407/165606975-0bea8041-b7f0-4518-8259-a992d1a3b061.gif)

3. Not a huge deal, but link highlighting does not work unless you manually type a valid prefix, e.g. `http://url`. Should we consider changing the default `url` placeholder that appears when you click the link button, so that it registers as a link on insertion?

4. All of the list item buttons don't appear to have working highlighting - is this intentional? (no worries if so, just checking)

constancecchen commented on src/components/markdown_editor/markdown_editor.spec.tsx:
```diff
@@ -0,0 +1,84 @@
+
+[link](https://elastic.co)
+
+!{tooltip[text](help)}`;
+      cy.realMount(
```
super minor newline nit, this tripped me up a bit at first glance
```suggestion
!{tooltip[text](help)}`;

      cy.realMount(
```

Alternatively I'd also say you could put `value` up on line 13 before the first describe() block if you don't want to deal with indentation shenanigans

chandlerprall replied:
added the newline 👍

constancecchen commented on src/components/markdown_editor/markdown_editor_toolbar.tsx:
```diff
@@ -88,6 +89,42 @@ const quoteCodeLinkButtons = [
   },
 ];
 
+interface FormatSelectableButtonArgs {
```
not sure if you think this is worth renaming since converting it to a component, I'm fine either way!

chandlerprall replied:
Definitely is, done & pushed

constancecchen reviewed (commented):
Code looks great to me! My only remaining comments are on the QA items above, particularly the quote button behavior

1Copenut commented on src/components/markdown_editor/markdown_editor_toolbar.tsx:
```diff
@@ -88,6 +89,42 @@ const quoteCodeLinkButtons = [
+      onClick={() => handleMdButtonClick(id)}
+      iconType={icon}
+      aria-label={label}
+      aria-pressed={isSelected}
```
I think I'm going to opt for removing the `aria-pressed` attribute. Reading through the [MDN docs on aria-pressed](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-pressed), I came across this line that gave me pause:

> Toggle buttons require a full press-and-release cycle to change their value. Pressing and releasing it once changes the value to `true`. If it's pressed and released again, the value changes back to `false`.

Our Markdown editor toggles between true and false based on the user's cursor position, so could cycle outside the full keypress and release cycle. Given this fact, I feel it has a reduced relationship with the user than say a Mute button for an audio player.

chandlerprall replied:
Thanks for taking a look! I removed the `aria-pressed` addition

1Copenut reviewed (changes requested):
I really wanted to keep the `aria-pressed` attribute, but came down on the side of removing it. Reasons and documentation linked in the review notes.

constancecchen commented:
jenkins test this

kibanamachine commented:
Preview documentation changes for this PR: https://eui.elastic.co/pr_5840/

chandlerprall commented:
@constancecchen 

> Quotes only appear to highlight when the first word of the quote, and not when clicking anywhere in the quote without selection

This happens because the `quote` node's content is a `paragraph`, and without addressing nested nodes it can't detect it is in a quote. That said, I've removed the quote button highlighting as clicking it when highlighted didn't toggle the quote syntax away, but instead inserted a second quote.

> Not a huge deal, but link highlighting does not work unless you manually type a valid prefix, e.g. http://url. Should we consider changing the default url placeholder that appears when you click the link button, so that it registers as a link on insertion?

Done! Updated the default url to `https://` which triggers the button association.

> All of the list item buttons don't appear to have working highlighting - is this intentional? (no worries if so, just checking)

Intentional, same reason that I gave for disabling the quote button.

chandlerprall commented:
@miukimiu pushed a fix for that bug

1Copenut reviewed (approved):
LGTM!

constancecchen reviewed (approved):
Fixes look great!

miukimiu reviewed (approved):
Thanks, @chandlerprall. Tested again and the error is gone.  LGTM! 🎉
